accepted/single-file/design.md

```diff
@@ -66,225 +32,254 @@ Publishing to a single file can be triggered by adding the following property to
 
-On Startup, the AppHost checks if it has embedded files. If so, it 
+## The Bundler
 
-* Memory maps the entire bundle file.
-* Extracts the necessary files (ex: native binaries) to disk (if any) as explained in this [document](extract.md). 
-* Sets up data-structures so other components can access files embedded directly, as explained in this [document](direct.md)
+The bundler is a tool that embeds the managed app and its dependencies into the native `AppHost` executable. The functional details of the bundler are explained in [this document](bundler.md). 
 
-#### Dependency Resolution
+## The Host
+
+### Startup
+
+On Startup, the [host components](https://github.com/dotnet/core-setup/blob/master/Documentation/design-docs/host-components.md) perform the following functions:
+
+* **AppHost**: The AppHost identifies itself as a single-file bundle (by checking the [bundle marker](bundler.md#bundle-marker)) before invoking [HostFxr](https://github.com/dotnet/core-setup/blob/master/Documentation/design-docs/host-components.md#host-fxr).
+
+* **HostFxr**: If invoked from a single-file app, HostFxr process the `runtimeconfig.json` and `deps.json` files directly from the bundle. The location of these `json` files  are identified directly from the bundle-header for simple access.
+
+* **HostPolicy**: Much of the bundle-processing is performed within HostPolicy.
+  
+  * If the app needs extraction, extracts out the appropriate files as explained in [this document](extractor.md).
+  
+  * Reads the `deps.json` file directly from the bundle and resolves dependencies.
+  
+  * Processes the Bundle manifest to maintain internal data-structures to locate bundled assemblies when probed by the runtime.
+  
+  * Implements a bundle-probe function, which will be used by the runtime to probe contents of the bundle when resolving assemblies. 
+  
+    * Bundle probe has the following signature:
+  
+      ```C++
+      /// <summary>
+      /// <param name="path"> Relative-path to the file being probed. </param>
+      /// <param name="size"> Out-param: size of the file, if found. </param>
+      /// <param name="path"> Out-param: offset within the bundle, if found</param>
+      /// <returns> true if the requested file is found in the bundle, 
+      ///           false otherwise. </returns>
+      /// </summary>
+      bool bundle_probe(const char *path, int64_t *size, int64_t *offset);
+      ```
+  
+    * The `bundle_probe` function pointer is passed to the runtime (encoded as a string) through a property named `BUNDLE_PROBE`.
+  
+    * The contents of the bundle are not included in `TRUSTED_PLATFORM_ASSEMBLIES`. Any additional assemblies on disk (ex: due to `additional-deps` /`AdditionalProbingPaths`) are included in `TRUSTED_PLATFORM_ASSEMBLIES`.
+  
```

Comment:
> "Tricks" for the few native dlls from CoreFX which will be bundled directly

We should get these linked into coreclr.dll to avoid need for these tricks. We are doing that for System.Globalization.Native already - https://github.com/dotnet/runtime/pull/6928

Reply:
The extraction directory (if any) will be added to `NATIVE_DLL_SEARCH_DIRECTORIES`.

`PLATFORM_RESOURCE_ROOTS` is treated similar to `TRUSTED_PLATFORM_ASSEMBLIES` -- the runtime probes the satellite assemblies in the bundle before looking elsewhere. I'll update the notes about it.

Reply:
By "tricks" I meant changes we made in the prototype so that `[DllImport("clrcompression.dll")]` continues to work even when `clrcompression.dll` is statically linked with the host. I don't think that will just magically work.
Another consideration is that `hostpolicy` is not aware of these changes, so maybe we will need to make it aware (so that it doesn't put a relative path to the `NATIVE_DLL_SEARCH_DIRECTORIES` as that may actually cause potential security issues).